kingpin/actors/rightscale/mci.py

```diff
@@ -0,0 +1,295 @@
+       cloud to launch. See below for details.
+
+    **Image Definitions**
+
+    Each cloud image definition is a dictionary that takes a few keys.
+
+    :cloud:
+      The name of the cloud as found in RightScale. We use the cloud 'Name'
+      which can be found in your `Settings -> Account Settings -> Clouds ->
+      insert_cloud_here` page. For example `AWS us-west-2`.
+
+    :image:
+      The cloud-specific Image UID. For example `ami-a1234abc`.
+
+    :instance_type:
+      The default instance type to launch when this AMI is launched. For
+      example, `m1.small`.
+      (*optional*)
+
+    :user_data:
+      The custom user data to pass to the instance on-bootup.
+      (*optional*)
+
+    **Examples**
+
+    .. code-block:: json
+
+        { "actor": "rightscale.mci.Create",
+          "desc": "Create an MCI",
+          "options": {
+              "name": "Ubuntu i386 14.04",
+              "description": "this is our test mci",
+              "images": [
+                {
+                    "cloud": "EC2 us-west-2",
+                    "image": "ami-e29774d1",
+                    "instance_type": "m1.small",
+                    "user_data": "cd /bin/bash"
+                },
+                {
+                    "cloud": "EC2 us-west-1",
+                    "image": "ami-b58142f1",
+                    "instance_type": "m1.small",
+                    "user_data": "cd /bin/bash"
+                }
+              ]
+          }
+        }
+
+    """
+
+    all_options = {
+        'name': (str, REQUIRED, 'The name of the MCI to be created.'),
+        'description': (
+            str, '',
+            'The description of the MCI to be created.'),
+        'images': (
+            list, [],
+            'A list of objects that describe our per cloud image settings.'),
+    }
+
+    def __init__(self, *args, **kwargs):
+        """Validate the user-supplied parameters at instantiation time."""
+
+        super(Create, self).__init__(*args, **kwargs)
+
+        allowed_image_options = (
+            'cloud', 'image', 'instance_type', 'user_data')
+        required_image_options = ('cloud', 'image')
+
+        for image in self.option('images'):
+            # Sanity check that no extra options were passed in
+            for key in image.keys():
+                if key not in allowed_image_options:
+                    raise exceptions.InvalidOptions(
+                        'Invalid option (%s) found in Image %s' % (key, image))
+
+            # Make sure that the required options were passed in
+            for required in required_image_options:
+                if required not in image.keys():
+                    raise exceptions.InvalidOptions(
+                        'Missing option "%s" in Image %s' %
+                        (required, image))
+
+    @gen.coroutine
+    def _get_image_def(self, description):
+        """Returns a fully populated set of Multi Cloud Image settings.
```

Comment:
nitpicky, but the docstring here is inconsistent about using `set` vs `array of tuples` when describing the return data.